velox/functions/prestosql/tests/ArrayFlattenTest.cpp

```diff
@@ -77,4 +77,169 @@ TEST_F(ArrayFlattenTest, nullArray) {
+
+  // [[null, 6, 5, null]]
+  // null
+  // [[3, 3, 1, 1]]
+  const auto expected = makeNullableArrayVector<int64_t>(
+      {{{std::nullopt, 6, 5, std::nullopt}}, std::nullopt, {{3, 3, 1, 1}}});
+
+  testExpression("flatten(c0)", {arrayOfArrays}, expected);
+}
+
+TEST_F(ArrayFlattenTest, constantArrayWithoutCopy) {
+  const auto baseVector = makeNullableArrayVector<int64_t>(
+      {{{1, 1}},
+       std::nullopt,
+       {{3, 3}},
+       {},
+       {{5, std::nullopt}},
+       {{std::nullopt, 6}}});
+  // Create arrays of array vector using above base vector.
+  // [[1, 1], null, [3, 3]]
+  // null
+  // [[], [5, null], [null, 6]]
+  const auto arrayOfArrays = makeArrayVector({0, 3, 3}, baseVector, {1});
+
+  // Create three element constant array by referencing the first element in
+  // 'arrayOfArrays'.
+  const auto constArrayOfArrays =
+      BaseVector::wrapInConstant(3, 0, arrayOfArrays);
+  // [[1, 1, 3, 3]]
+  // [[1, 1, 3, 3]]
+  // [[1, 1, 3, 3]]
+  const auto expected = makeArrayVector<int64_t>(
+      {{{1, 1, 3, 3}}, {{1, 1, 3, 3}}, {{1, 1, 3, 3}}});
+
+  // Enforce the constant encoding flatten path to trigger.
+  execCtx_.testingPeelingEnabled(false);
+  testExpression("flatten(c0)", {constArrayOfArrays}, expected);
+}
+
+TEST_F(ArrayFlattenTest, constantArrayWithCopy) {
+  const auto baseVector = makeNullableArrayVector<int64_t>(
+      {{{1, 1}},
+       std::nullopt,
+       {{3, 3}},
+       {},
+       {{5, std::nullopt}},
+       {{std::nullopt, 6}}});
+  // Create arrays of array vector using above base vector.
+  const auto numElements = 6;
+  auto reversedIndices = allocateIndices(numElements, pool());
+  auto* rawReversedIndices = reversedIndices->asMutable<vector_size_t>();
+  for (int i = 0; i < numElements; ++i) {
+    rawReversedIndices[i] = numElements - i - 1;
+  }
+  const auto mappedVector = wrapInDictionary(reversedIndices, baseVector);
+  // Create arrays of array vector using above base vector.
+  // [[null, 6], [5, null], []]
+  // null
+  // [[3, 3], null, [1, 1]]
+  const auto arrayOfArrays = makeArrayVector({0, 3, 3}, mappedVector, {1});
+
+  // Create three element constant array by referencing the first element in
+  // 'arrayOfArrays'.
+  const auto constArrayOfArrays =
+      BaseVector::wrapInConstant(3, 0, arrayOfArrays);
+  // [[null, 6, 5, null]
+  // [[null, 6, 5, null]
+  // [[null, 6, 5, null]
+  const auto expected = makeNullableArrayVector<int64_t>(
+      std::vector<std::vector<std::optional<int64_t>>>{
+          {std::nullopt, 6, 5, std::nullopt},
+          {std::nullopt, 6, 5, std::nullopt},
+          {std::nullopt, 6, 5, std::nullopt}});
+  for (bool enabled : {false, true}) {
+    SCOPED_TRACE(fmt::format("peeling encoding: {}", enabled));
+    // Enforce the constant encoding flatten path to trigger.
+    execCtx_.testingPeelingEnabled(enabled);
+    testExpression("flatten(c0)", {constArrayOfArrays}, expected);
+  }
+}
+
+TEST_F(ArrayFlattenTest, fuzz) {
```

Comment:
Do we need this? Let's just run fuzzer biased towards this function.

Reply:
I feel it might be good to have a dedicated fuzz test and catch some bug during the development.